Bluetooth: Controller: Fix assertion scheduling first BIG event and in window widening CIS event

cvinayak opened the pull request:
Fix assertion due to delayed prepare of first BIG event when
supporting encryption. Crypto calculation introduce extra
processing delay causing the scheduling of first BIG event
to be delayed. Detect such delay and skip to next ISO
interval.

Relax radio_tmr_start_us() delayed start assertion when
window widening in progress.

Fix Controller assertion handler to lock IRQs and spinlock
on assertions.

Fix uninitialized CIS LLL structure member variables,
uninitialised CIS acl_handle causes HCI command disallowed
for acl_handle's other than 0x0000.
Initialize other variables like sn, nesn etc. when create
CIS control procedure is performed.

Fixes #54642.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix uninitialized CIS LLL member variables

Fix uninitialized CIS LLL structure member variables,
uninitialised CIS acl_handle causes HCI command disallowed
for acl_handle's other than 0x0000.
Initialize other variables like sn, nesn etc. when create
CIS control procedure is performed.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Relax radio_tmr_start_us assert on win widening

Relax radio_tmr_start_us() delayed start assertion when
window widening in progress.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix radio_tmr_aa_save() value for ISO receive

Fix the access address and radio ready timestamp to derive
the value of the first subevent, so that correct anchor
point reference is used in scheduling subsequent subevents.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix assertion scheduling first BIG event

Fix assertion due to delayed prepare of first BIG event when
supporting encryption. Crypto calculation introduce extra
processing delay causing the scheduling of first BIG event
to be delayed. Detect such delay and skip to next ISO
interval.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: hci_rpmsg: Lock IRQs and spinlock on assertions

Fix Controller assertion handler to lock IRQs and spinlock
on assertions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Schedule BIG event after Periodic Advertising

Add implementation to schedule BIG events after Periodic
Advertising when Periodic Advertising is enabled after
Extended Advertising has already been enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix GPIO debug pins for ISO Central

Fix the GPIO debug pin assignment for ISO central role.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Refactor duplicate aux_time_get implementation

Refactor duplicate aux_time_get() implementation and reuse
the same code.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_central_iso.c:
```diff
@@ -696,6 +697,13 @@ uint8_t ull_central_iso_setup(uint16_t cis_handle,
 	cis->offset = cis_offset;
 	cis->central.instant = instant;
 	cis->lll.event_count = -1;
+	cis->lll.next_subevent = 0;
+	cis->lll.sn = 0;
+	cis->lll.nesn = 0;
+	cis->lll.cie = 0;
+	cis->lll.flushed = 0;
+	cis->lll.active = 0;
+	cis->lll.datapath_ready_rx = 0;
```
Should all these be `0U`?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_peripheral_iso.c:
```diff
@@ -316,6 +309,13 @@ uint8_t ull_peripheral_iso_setup(struct pdu_data_llctrl_cis_ind *ind,
 	cis->offset = sys_get_le24(ind->cis_offset);
 	memcpy(cis->lll.access_addr, ind->aa, sizeof(ind->aa));
 	cis->lll.event_count = -1;
+	cis->lll.next_subevent = 0;
+	cis->lll.sn = 0;
+	cis->lll.nesn = 0;
+	cis->lll.cie = 0;
+	cis->lll.flushed = 0;
+	cis->lll.active = 0;
+	cis->lll.datapath_ready_rx = 0;
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -521,6 +524,22 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
+	/* Check and skip to next interval if CPU processing introduces latency
+	 * that can delay scheduling the first ISO event.
+	 */
+	ticks_expire = ftr->ticks_anchor - ticks_slot_offset +
+		       HAL_TICKER_US_TO_TICKS(sync_iso_offset_us);
+	ticks_threshold = ticker_ticks_now_get() +
+			  HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_START_US);
+	ticks_diff = ticker_ticks_diff_get(ticks_expire, ticks_threshold);
+	if (ticks_diff & BIT(HAL_TICKER_CNTR_MSBIT)) {
+		sync_iso_offset_us += interval_us -
+			lll->window_widening_periodic_us;
+		lll->window_widening_event_us +=
+			lll->window_widening_periodic_us;
+		lll->payload_count += lll->bn;
+	}
```
Any risks of underflow or overflow on any of these?

cvinayak replied:
`ticks_*` are permited to underflow and overflow as it is free-running. `window_widening_periodic_us` is a fraction of `interval_us` hence would not lead to negative offset.

Thalley reviewed (dismissed):
Looks OK to me, with a few comments

cvinayak committed:
Bluetooth: Controller: Fix auxiliary PDU time reservation calculation

Fix auxiliary PDU time reservation calculation to reserve
for possible increase in PDU len due to updates to data in
the Common Extended Header Format of the PDUs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor rename to sync_time_get()

Minor rename of file static function sync_time_get().

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Periodic Advertising time reservation

Fix Periodic Advertising time reservation when enabling with
ADI support. As the PDU is updated, use the updated PDU to
calculate the time reservation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use context specific get interface

As the range of ticker_id has been checked, context specific
get interface be used instead of doing range check again in
ull_hdr_get_cb().

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -522,19 +522,30 @@ static void isr_rx(void *param)
+
 	/* Save the AA captured for the first anchor point sync */
 	if (!radio_tmr_aa_restore()) {
+		struct lll_sync_iso_stream *stream;
```
```suggestion
		const struct lll_sync_iso_stream *stream;
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1509,8 +1511,11 @@ uint8_t ll_adv_enable(uint8_t enable)
 						EVENT_OVERHEAD_START_US +
 						(EVENT_TICKER_RES_MARGIN_US << 1));
 
+				ticks_slot_overhead =
+					ull_adv_sync_evt_init(adv, sync, NULL);
```
```suggestion
				ticks_slot_overhead = ull_adv_sync_evt_init(adv, sync, NULL);
```

The TSC decided this week to remove the 80 character suggestion, so now we can (and should) use the 100 character limit instead.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -868,15 +880,28 @@ uint8_t ll_adv_sync_enable(uint8_t handle, uint8_t enable)
 				HAL_TICKER_US_TO_TICKS(EVENT_OVERHEAD_START_US);
+
+#if defined(CONFIG_BT_CTLR_SCHED_ADVANCED)
+			err = ull_sched_adv_aux_sync_free_slot_get(
+							TICKER_USER_ID_THREAD,
+							sync->ull.ticks_slot,
+							&ticks_anchor_sync);
```
err = ull_sched_adv_aux_sync_free_slot_get(TICKER_USER_ID_THREAD,
								   sync->ull.ticks_slot,
								   &ticks_anchor_sync);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -1053,26 +1079,23 @@ uint32_t ull_adv_sync_time_get(const struct ll_adv_sync_set *sync,
+		const struct lll_adv_sync *lll_sync = &sync->lll;
 
-	time_us = ull_adv_sync_pdu_time_get(sync, ter_pdu);
+		pdu = lll_adv_sync_data_peek(lll_sync, NULL);
```
This could probably just be

```suggestion
		pdu = lll_adv_sync_data_peek(&sync->lll, NULL);
```

without loosing readability

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -1888,7 +1924,7 @@ static inline uint16_t sync_handle_get(struct ll_adv_sync_set *sync)
 }
 
-static uint32_t ull_adv_sync_pdu_time_get(const struct ll_adv_sync_set *sync,
+static uint32_t sync_time_get(const struct ll_adv_sync_set *sync,
 					  const struct pdu_adv *pdu)
```
Fix indentation

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_central_iso.c:
```diff
@@ -504,7 +504,7 @@ uint8_t ll_cis_create_check(uint16_t cis_handle, uint16_t acl_handle)
 		/* Verify handle validity and association */
 		cis = ll_conn_iso_stream_get(cis_handle);
-		if (cis->lll.handle == cis_handle && cis->lll.acl_handle == acl_handle) {
+		if (cis->lll.handle == cis_handle) {
```
Just to make sure: The `handle`/`cis_handle` here are the connection handle, and not the CIS ID, right?

cvinayak replied:
these are connection handles, not CIS ID.

Thalley reviewed (changes requested):
A few minor comments
